core/server/src/main/java/alluxio/master/file/meta/InodeDirectory.java

```diff
@@ -76,15 +77,20 @@ protected InodeDirectory getThis() {
    * @param child the inode to add
    */
   public void addChild(Inode<?> child) {
-    mChildren.add(child);
+    synchronized (child) {
```

Comment:
I think this is one of the problems with using index directly. It makes users of the index much more complicated and error prone.